src/storage/txn/scheduler.rs

```diff
@@ -918,46 +958,134 @@ impl Scheduler {
 
     /// Initiates an async operation to get a snapshot from the storage engine, then posts a
     /// `SnapshotFinished` message back to the event loop when it finishes.
-    fn get_snapshot(&mut self, cid: u64) {
-        SCHED_STAGE_COUNTER_VEC.with_label_values(&[self.get_ctx_tag(cid), "snapshot"]).inc();
+    fn get_snapshot(&mut self, ctx: &Context, cids: Vec<u64>) {
+        for cid in &cids {
+            SCHED_STAGE_COUNTER_VEC.with_label_values(&[self.get_ctx_tag(*cid), "snapshot"]).inc();
+        }
+        let cids1 = cids.clone();
         let ch = self.schedch.clone();
         let cb = box move |(cb_ctx, snapshot)| {
             match ch.send(Msg::SnapshotFinished {
-                cid: cid,
+                cids: cids1,
                 cb_ctx: cb_ctx,
                 snapshot: snapshot,
             }) {
                 Ok(_) => {}
                 e @ Err(TransportError::Closed) => info!("channel closed, err {:?}", e),
-                Err(e) => panic!("send SnapshotFinish failed cmd id {}, err {:?}", cid, e),
+                Err(e) => panic!("send SnapshotFinish failed, err {:?}", e),
             }
         };
 
-        if let Err(e) = self.engine.async_snapshot(self.extract_context(cid), cb) {
-            SCHED_STAGE_COUNTER_VEC.with_label_values(&[self.get_ctx_tag(cid), "async_snap_err"])
-                .inc();
-            self.finish_with_err(cid, Error::from(e));
+        if let Err(e) = self.engine.async_snapshot(ctx, cb) {
+            for cid in cids {
+                SCHED_STAGE_COUNTER_VEC
+                    .with_label_values(&[self.get_ctx_tag(cid), "async_snap_err"])
+                    .inc();
+
+                let e = e.maybe_clone().unwrap_or_else(|| {
+                    error!("async snapshot failed for cid={}, error {:?}", cid, e);
+                    EngineError::Other(box_err!("{:?}", e))
+                });
+                self.finish_with_err(cid, Error::from(e));
+            }
+        }
+    }
+
+    /// Initiates an async operation to batch get snapshot from the storage engine, then posts a
+    /// `BatchSnapshotFinished` message back to the event loop when it finishes, also it posts a
+    /// `ReadIndexSnapshot` message if there are any `None` responses.
+    fn batch_get_snapshot(&mut self, batch: Vec<(Context, Vec<u64>)>) {
+        let mut all_cids = Vec::with_capacity(batch.iter()
+            .fold(0, |acc, &(_, ref cids)| acc + cids.len()));
+        for &(_, ref cids) in &batch {
+            all_cids.extend(cids);
+        }
+
+        let batch1 = batch.clone();
+        let ch = self.schedch.clone();
+        let on_finished: engine::BatchCallback<Box<Snapshot>> = box move |results: Vec<_>| {
+            let mut ready = Vec::with_capacity(results.len());
+            let mut ready_ids: Vec<u64> = Vec::new();
+            let mut backlog = Vec::new();
```

Comment:
rename backlog.

ReadIndex is not proper here, the flow here is 

1. Get snapshot in batch
2. For some regions if we can't get the snapshot because of leader lease expired, we must re-get snapshot again.

For the re-get snapshot, if the leader is in the lease again, we don't need to call ReadIndex but can get the local snapshot. So here GetSnapshot or ReGetSnapshot may be better.